examples/tc_neighbor_sharing.py

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python
+# Copyright (c) PLUMgrid, Inc.
+# Licensed under the Apache License, Version 2.0 (the "License")
+
+# This example shows how a combination of BPF programs can be used to perform
+# per-IP classification and rate limiting. The simulation in this example
+# shows an example where N+M devices are combined and use 1 WAN. Traffic sent
+# from/to the "neighbor" devices have their combined bandwidth capped at
+# 128kbit, and the rest of the traffic can use an additional 1Mbit.
+
+# This works by sharing a map between various tc ingress filters, each with
+# a related set of bpf functions attached. The map stores a list of dynamically
+# learned ip addresses that were seen on the neighbor devices and should be
+# throttled.
+
+#                          /------------\                        |
+# neigh1 --|->->->->->->->-|            |                        |
+# neigh2 --|->->->->->->->-|    <-128kb-|        /------\        |
+# neigh3 --|->->->->->->->-|            |  wan0  | wan  |        |
+#          | ^             |   br100    |-<-<-<--| sim  |        |
+#          | clsfy_neigh() |            |   ^    \------/        |
+# lan1 ----|->->->->->->->-|    <--1Mb--|   |                    |
+# lan2 ----|->->->->->->->-|            |   classify_wan()       |
+#            ^             \------------/                        |
+#            pass()                                              |
+
+
+from bpf import BPF
+from pyroute2 import IPRoute, NetNS, IPDB, NSPopen
+import sys
+from time import sleep
+
+ipr = IPRoute()
+ipdb = IPDB(nl=ipr)
+b = BPF(src_file="tc_neighbor_sharing.c", debug=0)
+
+wan_fn = b.load_func("classify_wan", BPF.SCHED_CLS)
+pass_fn = b.load_func("pass", BPF.SCHED_CLS)
+neighbor_fn = b.load_func("classify_neighbor", BPF.SCHED_CLS)
+
+num_neighbors = 3
+num_locals = 2
+
+# class to build the simulation network
+class SharedNetSimulation(object):
+
+    def __init__(self):
+        self.ipdbs = []
+        self.namespaces = []
+        self.processes = []
+
+        # Create the wan namespace, and attach an ingress filter for throttling
+        # inbound (download) traffic
+        (self.wan, wan_if) = self._create_ns("wan0", "172.16.1.5/24", None)
+        ipr.tc("add", "ingress", wan_if["index"], "ffff:")
+        ipr.tc("add-filter", "bpf", wan_if["index"], ":1", fd=wan_fn.fd,
+               prio=1, name=wan_fn.name, parent="ffff:", action="drop",
+               classid=1, rate="128kbit", burst=1024 * 32, mtu=16 * 1024)
+        ipr.tc("add-filter", "bpf", wan_if["index"], ":2", fd=pass_fn.fd,
+               prio=2, name=pass_fn.name, parent="ffff:", action="drop",
+               classid=2, rate="1024kbit", burst=1024 * 32, mtu=16 * 1024)
+        self.wan_if = wan_if
+
+    # helper function to create a namespace and a veth connecting it
+    def _create_ns(self, name, ipaddr, fn):
```

Comment:
Not that easily, unfortunately, the vlan example is doing some things with mac addresses that makes it less portable.